apps/jupyter/extensions/notebook-runner/src/iframe-api.ts

```diff
@@ -407,6 +407,67 @@ export class EmbeddedPythonCallbacks {
       throw new Error(`Unsupported content format: ${file.format}`);
     }
   }
+
+  private async writeFolderContents(
+    basePath: string,
+    files: Map<string, Blob>,
+  ): Promise<void> {
+    for (const [relativePath, blob] of files.entries()) {
+      const path = `${basePath}/${relativePath}`;
+      const pathParts = relativePath.split("/");
+
+      if (pathParts.length > 1) {
+        let currentPath = basePath;
+
+        for (let i = 0; i < pathParts.length - 1; i++) {
+          currentPath += `/${pathParts[i]}`;
+
+          try {
+            await this.createFolder(currentPath, pathParts[i]);
+          } catch {
+            // folder may already exist, ignore errors
+          }
+        }
+      }
+      await this.putFileContents(path, blob);
+    }
+  }
+
+  private async readFolderContents(
+    basePath: string,
+  ): Promise<Map<string, Blob>> {
+    const files = new Map<string, Blob>();
+
+    try {
+      const folder = await this.app.serviceManager.contents.get(basePath, {
+        content: true,
+      });
+
+      if (folder.type !== "directory") {
+        return files;
```

Comment:
It seems we probably want to throw here, no? Trying to call `readFolderContents` on a file is probably not something we want to support or am I missing something?